### PR TITLE
Language selection

### PR DESCRIPTION
**Pull request by eorahil**
Added language selection buttons after importing MoM
Stored languages lists and current user language in config.ini file
modified Textbutton and RectangleBorder clases in order to be able to change it's color dinamically.

### CODE REVIEW

**Comment by NPBruce** [comment]
The extract ini files are for internal use and are generated, as such they shouldn't be added to source control, please remove.

File: Assets/Scripts/Game.cs [hunk 1]
```diff
@@ -72,6 +75,13 @@ void Start()
         // Create some things
         uiScaler = new UIScaler(uICanvas);
         config = new ConfigFile();
+        currentLang = 1; // English
+
+        if (config.data.Get("MoMConfig") != null)
```
Comment (NPBruce) [inline Assets/Scripts/Game.cs]: Better to use:

if (config.data.Get("MoMConfig", "currentLang") != "")

This allows for the config to exist but language not set (as would be for existing users).

File: Assets/Scripts/GameSelection.cs [hunk 2]
```diff
@@ -1,13 +1,17 @@
-﻿using UnityEngine;
-using System.Collections;
+﻿using System;
```
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: Haven't tested this but I am avoiding using System as there are many conflicts between System and UnityEngine.  Where I need system I just specify explicitly.

**Comment by NPBruce** [comment]
Just did a super quick pass, will have a look at this properly when I get some time.

File: Assets/Scripts/Game.cs (hunk 1 above)
Comment (NPBruce) [inline Assets/Scripts/Game.cs]: Actually, why is the language in MoMConfig anyway?  I would expect all game types to share language settings.

File: Assets/Scripts/GameSelection.cs [hunk 3]
```diff
@@ -81,30 +79,91 @@ public GameSelection()
...
+        {
+            if (game.config.data.Get("MoMConfig") == null)
             {
-                tb = new TextButton(new Vector2((UIScaler.GetWidthUnits() - 10) / 2, 23.2f), new Vector2(10, 2f), "Import Content", delegate { Import("MoM"); });
-                tb.background.GetComponent<UnityEngine.UI.Image>().color = new Color(0, 0.03f, 0f);
+                // TODO: FFGlookup to get the first line (languages list)
+                string langsString = StringI18n.FFG_LANGS;
+
+                // List of languages
+                game.config.data.Add("MoMConfig","Langs", langsString);
```
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: Why are we saving available languages as a config item?

Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: And why is this dependant on if MoM doesn't need an import?

Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: If the reason we save the list of strings is so that the index we save will still work if the list changes order, wouldn't it be better to just save the string name of the language?

File: Assets/Scripts/GameSelection.cs [hunk 4]
```diff
@@ -81,30 +79,91 @@ public GameSelection()
...
+                game.config.Save();
             }
-            else // Import OK, can redo
+
+            string[] langs = game.config.data.Get("MoMConfig", "Langs").Split(',');
+            int currentLang = int.Parse(game.config.data.Get("MoMConfig", "currentLang"));
+
+            game.currentLang = currentLang;
+
+            float verticalStart = 12 - ((langs.Length - 1) * 1.25f);
```
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: any reason this isn't:

float verticalStart = 15 - (langs.Length * 1.25f);

Or

float verticalStart = UIScaler.GetVCenter(langs.Length * -1.25f);


For vertical center?

File: Assets/Scripts/GameSelection.cs [hunk 5]
```diff
@@ -81,30 +79,91 @@ public GameSelection()
...
+            languageTextButtons = new TextButton[langs.Length];
+            for (int i = 1; i < langs.Length; i++)
             {
-                tb = new TextButton(new Vector2((UIScaler.GetWidthUnits() - 10) / 2, 23.2f), new Vector2(10, 2f), "Reimport Content", delegate { Import("MoM"); });
-                tb.background.GetComponent<UnityEngine.UI.Image>().color = new Color(0, 0.03f, 0f);
+                // Need current index in order to delegate not point to loop for variable
+                int currentIndex = i;
+
+                languageTextButtons[currentIndex] = new TextButton(
+                    new Vector2(UIScaler.GetHCenter() + 15, verticalStart + (2.5f * currentIndex)),
```
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: Using GetRight(-10) is probably a better approach.  Having said that both of these won't work on 4x3 screens (as written will go over the edge, getright will overlap with game type buttons.

Given that at some point we will need audio settings (and maybe others) creating a settings page from the main menu is one option.  Could also put a button to it here on the main splash screen.

File: Assets/Scripts/GameSelection.cs [hunk 6]
```diff
@@ -81,30 +79,91 @@ public GameSelection()
...
+            languageTextButtons[currentGame.currentLang].setColor(Color.gray);
         }
-        new TextButton(new Vector2(1, UIScaler.GetBottom(-3)), new Vector2(8, 2), "Exit", delegate { Exit(); }, Color.red);
+        languageTextButtons[i].setColor(Color.white);
+        // English is the default current language
+        currentGame.config.data.Add("MoMConfig", "currentLang", i.ToString());
+        currentGame.config.Save();
+        currentGame.currentLang = i;
+        refreshScreen();
+        Debug.Log("new current language stablished:" + i.ToString());
```
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: Strange unity quirk: For logs like this add System.Environment.NewLine to the end, or it will put stack trace info into what should be just an informative entry.

File: Assets/Scripts/Content/StringI18n.cs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+                List<String> finalTranslation = new List<string>();
+                String currentTranslation = "";
+                bool opened = false;
+                foreach (String suposedTranslation in partialTranslation)
+                {
+                    currentTranslation += suposedTranslation;
+
+                    bool initialQuote = suposedTranslation.Length != 0 && suposedTranslation[0] == QUOTES;
+                    bool finalQuote = suposedTranslation.Length > 1 &&
+                        suposedTranslation[suposedTranslation.Length - 1] == QUOTES;
```
Comment (NPBruce) [inline Assets/Scripts/Content/StringI18n.cs]: This method may not work because \" is escaped by using "" in the csv.  This means that:

"My text has a ""quote"", and other things"

Won't work right.

**Comment by NPBruce** [comment]
Thanks for this, sorry if I am being too critical.  I'm sure much of my code is far worse!

File: Assets/Scripts/Game.cs (hunk 1 above)
Comment (eorahil) [inline Assets/Scripts/Game.cs]: I was confused about D2E having multilanguage, the Steam App doesn't have it. That was why I created an specific MoM config to use multilanguage but the lang selection only appears after MoM importing, not D2E. 
FFG texts for D2E arent in all languages, only english. So I decided to, in a first approach, only implement multilang for MoM.

if (config.data.Get("MoMConfig") != null). There is no such [MoMConfig] tag in previous version of Valkyrie. I added in this resquest. So What I'm looking for is exactly that. Previous users won't have [MoMConfig] tag and I will add it.

Comment (eorahil) [inline Assets/Scripts/Game.cs]: I will remove the extract-ini.files , shouldn't it be in the .gitignore file in order to not be included in git branches?

File: Assets/Scripts/GameSelection.cs (hunk 3 above)
Comment (eorahil) [inline Assets/Scripts/GameSelection.cs]: My first thoughts were about storing available languages. Initially there are the FFG ones, but later there will be the Valkyrie ones. Or even 2 lists.
There are the Valkyrie editor languages, and the scenario languages. A Italian guy can create (or translate) an scenario to italian but the valkyrie editor won't have the Italian language initially son he will create the Italian scenario using English Valkyrie.
So there should be 2 languages lists.
* The FFG one: The languages used by FFG. By reimporting, if FFG increases the languages list, we will adapt.
* The Valkyrie one: The languages which tranlation is available for Valkyrie editor. If you select a non Valkyrie language, the editor will be loaded in English.

There is no FFG D2E languages, so we could have even a valkyrie editor language (spanish) which won't be in FFG list.

File: Assets/Scripts/GameSelection.cs (hunk 4 above)
Comment (eorahil) [inline Assets/Scripts/GameSelection.cs]: The language 0 is the key (.) so the real list of languages will be the array size minus 1. 
Not exactly the vertical center but the starting point of a list of buttons which will be vertically centered. I we have only one button, the vertivalStart will be almost the center. But if we have 3 buttons, the VerticalStart will be in the point that, after drawing the 3 buttons, the midle one will be in the center.

File: Assets/Scripts/GameSelection.cs (hunk 5 above)
Comment (eorahil) [inline Assets/Scripts/GameSelection.cs]: Tryed the getRight option but in Unity, debugging the scene, i was far right and I wasn't visible. There is some strange behaviour with UIScaler. I think when 2 monitors are used, the screen width is calculated adding the width of both screens. This problem is only in Unity debug, but if I build an exe file the problem disappears.

Agree with the standalone settings page.

File: Assets/Scripts/Content/StringI18n.cs (hunk 7 above)
Comment (eorahil) [inline Assets/Scripts/Content/StringI18n.cs]: Ok, as soon as there was no double quotes before comma in FFG texts I didn't fix this but  I'm figuring out a solution counting number of quotes.

File: Assets/Scripts/Game.cs (hunk 1 above)
Comment (NPBruce) [inline Assets/Scripts/Game.cs]: I get that D2E doesn't have translations now, but I don't see why support shouldn't be common.  The format is the same and is still has this:

```
KEY,English,Spanish,French,German,Italian,Portuguese,Polish,Japanese,Chinese,Czech
```

If we went to exclude language options because we know they aren't actually available we could either check a test key (ie: CAMPAIGN_1_TITLE) or maybe even check all keys to see if that language has anything.

Either way I don't particularly like language selection being under "MoMConfig", even if for now we don't implemented it for any other game type.

tldr; please move language settings out of 'MoMConfig', but you can leave the rest if you want.

As for the extra files, even better would be to put them in a sensible place!  I have fixed this in development.

File: Assets/Scripts/GameSelection.cs (hunk 3 above)
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: The way I see it there are 3 lists:

* FFG
* Valkyrie
* Scenario/quest

I don't think there should be any need for any of these to have a consistent order/contents.  If they all start with a header line that tells you want languages they provide, they can all be different and unique.  Any change from FFG (ie: additional language) would also 'just work'.

Does that make sense?

File: Assets/Scripts/GameSelection.cs (hunk 4 above)
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: Yep, your are right about the -1, but I'm still confused about the '12'.

File: Assets/Scripts/GameSelection.cs (hunk 5 above)
Comment (NPBruce) [inline Assets/Scripts/GameSelection.cs]: That is odd, I use 2 screens and haven't seen this.  I use GetRight all over the place, do you have issues with other UI items?

In the play window in Unity you can select the aspect ratio to run, perhaps try picking different ratios and seeing what happens?

File: Assets/Scripts/Content/StringI18n.cs (hunk 7 above)
Comment (NPBruce) [inline Assets/Scripts/Content/StringI18n.cs]: This definitely occurs in D2E, and as mentioned above I would prefer a generic solution.

File: Assets/Scripts/GameSelection.cs (hunk 3 above)
Comment (eorahil) [inline Assets/Scripts/GameSelection.cs]: Ok, my intention was to use one single languages list in order to respect the array position for every language, but I can figure out a way to use the language name string instead of its position in order to adapt for every localization file.
I will rework it.

**Comment by eorahil** [comment]
I will rework it all, as soon as I have something I will commit and post comments.

**Comment by eorahil** [comment]
I'm commiting because I use different computers but don't check the changes until y post "finished" because it's still WIP.

**Comment by eorahil** [comment]
Ok changes made:
* Use [UserConfig] to Store language, in String format, not numeric.
* Created DictI18n class , a container for StringI18n classes.
* Used the D2E language list (including Czeck)
* Excluded in .gitignore the extract-*.ini files
* Moved language selection to a new Options Screen. Buttons are Centered (fixed my screen problems changing aspect)
* Added System.Environment.NewLine to all log traces
* The StringI18n quoted/comma transformations isn't fixed yet, but next step in translation will be the use of this system and it will be fixed and tested completely.
* Renamed and relocated the Screens of the App. I was kind of confused about what each class are. I think it's more clear to have all game Screens in same namespace (and folder) and adding a suffix "Screen".

Take a look and tell me

File: .gitignore [hunk 8]
```diff
@@ -33,3 +33,5 @@ sysinfo.txt
 # FFG imported content
 /content/D2E/ffg
 /content/MoM/ffg
+# Valkyrie ini files importing ffg ones
```
Comment (NPBruce) [inline .gitignore]: If you merge back from development I have moved these files, this change is not required.

**Comment by NPBruce** [comment]
Great, thanks.  Will test it as soon as I get a chance.

File: .gitignore (hunk 8 above)
Comment (eorahil) [inline .gitignore]: Merged and removed ignore

**Comment by NPBruce** [comment]
Great work!
